### PR TITLE
feat(db): add utf8mb4 support

### PR DESCRIPTION
**Pull request by vladikoff**
Connects to https://github.com/mozilla/fxa-auth-db-mysql/issues/229

### CODE REVIEW

**Comment by vladikoff** [comment]
This is missing `ALTER DATABASE fxa CHARACTER SET utf8mb4 COLLATE utf8mb4_bin`

File: lib/db/schema/patch-060-061.sql [hunk 1]
```diff
@@ -0,0 +1,61 @@
+--  Update devices table to utf8mb4
+SET NAMES utf8mb4 COLLATE utf8mb4_bin;
+
+ALTER DATABASE fxa CHARACTER SET utf8mb4 COLLATE utf8mb4_bin;
```
Comment (vladikoff) [inline lib/db/schema/patch-060-061.sql]: ⚠️

**Comment by jbuck** [comment]
It took 36 minutes to migrate a clone of production without prod traffic hitting it.

File: lib/db/mysql.js [hunk 2]
```diff
@@ -1013,10 +1027,20 @@ module.exports = function (log, error) {
...
-    return retryable(
-      this.getClusterConnection.bind(this, name),
-      [ER_TOO_MANY_CONNECTIONS, 'ECONNREFUSED', 'ETIMEDOUT', 'ECONNRESET']
-    )
+    return new P((resolve, reject) => {
+      retryable(
+        this.getClusterConnection.bind(this, name),
+        [ER_TOO_MANY_CONNECTIONS, 'ECONNREFUSED', 'ETIMEDOUT', 'ECONNRESET']
+      ).then((connection) => {
+        connection.query('SET NAMES utf8mb4 COLLATE utf8mb4_bin;', (err) => {
```
Comment (rfk) [inline lib/db/mysql.js]: IIUC this will execute "`SET NAMES`" for every db query, even if an existing connection is being re-used from the pool.  Over in profile-server we did a gross little hack to avoid the overhead here and only "`SET NAMES`" once per connection, it may be worth doing the same here:

  https://github.com/mozilla/fxa-profile-server/blob/master/lib/db/mysql/index.js#L236

Comment (rfk) [inline lib/db/mysql.js]: (We should also copy across the strict-mode enforcement from that repo, but we can do that as a separate PR)

**Comment by jbuck** [comment]
Update: ALGORITHM=INPLACE is not supported. Reason: Cannot change column type INPLACE. Try ALGORITHM=COPY.

Aurora only took 20 minutes to migrate the column though 🙃

**Comment by vladikoff** [comment]
from mtg: 

## New plan for train-94.1

* [ ] New Column `nameutf8`, default null col.
* [ ] New Procedure for updateDevices, writes to both.
* [ ] Always read the old column (no changes)

## After 94.1 deployed

* [ ] Only copy over names if it they are not null. Check in the migration script without LIMIT.
* [ ] Run the prod migration script with a limit

## train-95

* [ ] Replace the new procedure with a new one that reads from new column  
* [ ] Stop writing / reading to the old column

## train-95.1

* [ ] Leave the route alone until we start reading from the new column.

**Comment by rfk** [comment]
Thanks Vlad!  Adding some more notes for my own sanity-checking.  The important invariants of this procedure are:

* After v94.1 is deployed, all new device updates will write good utf8mb4 values into the new column.  We should test this as part of the deploy, by setting a device name to e.g. the CJK test strings from [1] ad checking that they make it into the db intact.

* After the manual update and before v95 is deployed, the new column will contain a complete record of everyone's device name, likely including literal "??" characters for users who hit this bug in production.

* After v95 is deployed, no stored procedures will touch the old column, rendering it safe to drop or rename or whatever.

Sounds 👍 to me.

[1] https://bugzilla.mozilla.org/show_bug.cgi?id=1348298

**Comment by vladikoff** [comment]
> all new device updates will write good utf8mb4 values into the new column. 

@rfk "write good" -> to clarify in my mind this means: "they will write values as utf8mb4 into the column". However the auth-server is not enabled to accept those values for that deploy.

**Comment by vladikoff** [comment]
> users who hit this bug in production.

@rfk  I assume this means they hit this bug before we enabled the emoji regex filter

File: test/local/mysql_tests.js [hunk 3]
```diff
@@ -345,6 +361,64 @@ describe('MySQL', () => {
...
+          }
+        )
+    }
+
+    return P.all([
+      checkDeviceName('name 🍓'),
+      checkDeviceName('User 在 home 上的 Nightly'),
+      checkDeviceName('𝌆 cool name'),
+      checkDeviceName('advanced emoji ✋🏼'),
+      checkDeviceName('Й,К,Л,М,Н,О,П,Р,С,Т,У,Ф,Х,Ц,Ч,Ш,Щ,Ъ,Ы,Ь,Э,Ю,Я'),
```
Comment (vladikoff) [inline test/local/mysql_tests.js]: @rfk feel free to add more exciting tests here :)

**Comment by vladikoff** [comment]
@rfk, @jbuck r? 

Are we missing anything for v94.1? Depending on the timing I might not able to make changes to this patch in time.

File: lib/db/mysql.js [hunk 4]
```diff
@@ -1015,10 +1031,28 @@ module.exports = function (log, error) {
...
+      retryable(
+        this.getClusterConnection.bind(this, name),
+        [ER_TOO_MANY_CONNECTIONS, 'ECONNREFUSED', 'ETIMEDOUT', 'ECONNRESET']
+      ).then((connection) => {
+
+        if (connection._fxa_initialized) {
+          return resolve(connection)
+        }
+
+        connection.query('SET NAMES utf8mb4 COLLATE utf8mb4_bin;', (err) => {
```
Comment (vladikoff) [inline lib/db/mysql.js]: @rfk Do we need a new issue filed to enforce strict mode? (separate PR)

**Comment by vladikoff** [comment]
One thing to keep in mind, in my Sequel Pro Tool I still get question marks for some emojis, I think it is a bug in that software: 

ar

**Comment by vladikoff** [comment]
@rfk to see MySql procedures / queries create something like this (change user dir): 

```
➜  ~ cat .my.cnf 
[mysqld]
general_log_file        = /Users/vladikoff/tmp/mysql.log
general_log             = 1
```

Restart MySQL server and then you should be able to see the queries:

```
➜  ~ tail -f ~/tmp/mysql.log
		    7 Quit	
		    8 Quit	
		    9 Quit	
		   10 Quit	
170822 12:15:35	   11 Connect	root@localhost on fxa
		   12 Connect	root@localhost on fxa
		   12 Query	SET NAMES utf8mb4 COLLATE utf8mb4_bin
		   11 Query	SET NAMES utf8mb4 COLLATE utf8mb4_bin
		   12 Query	CALL fetchVerificationReminders_2(1503418535556, 'second', '604800000', '1080000000', '20')
		   11 Query	CALL fetchVerificationReminders_2(1503418535554, 'first', '172800000', '601200000', '20')
```

**Comment by rfk** [comment]
> @rfk "write good" -> to clarify in my mind this means: "they will write values as utf8mb4 into the column". 

Right, exactly.

> However the auth-server is not enabled to accept those values for that deploy.

The auth-server won't accept emoji, but it *should* start accepting CJK characters as described in [1], which will give us something we can test with.

[1] https://bugzilla.mozilla.org/show_bug.cgi?id=1348298

**Comment by rfk** [comment]
>  I assume this means they hit this bug before we enabled the emoji regex filter

As above, I mean "tried to include CJK characters in their device name, and had them corrupted into question marks".

File: lib/db/mysql.js [hunk 5]
```diff
@@ -112,31 +124,33 @@ module.exports = function (log, error) {
...
-    return mysql.readFirstResult(DB_METADATA, [options.patchKey])
-      .then(
-        function (result) {
-          mysql.patchLevel = +result.value
-
-          log.info('connect', {
-            patchLevel: mysql.patchLevel,
-            patchLevelRequired: patch.level
-          })
+    return new P.resolve().then(() => {
```
Comment (rfk) [inline lib/db/mysql.js]: The use of `new P.resolve()` here seems odd; is it necessary at all, and if so should we remove the `new`?

File: test/local/mysql_tests.js (hunk 3 above)
Comment (rfk) [inline test/local/mysql_tests.js]: They look pretty comprehensive to me :-)

File: test/local/mysql_tests.js [hunk 6]
```diff
@@ -345,6 +361,64 @@ describe('MySQL', () => {
...
+    ]).catch(function (error) {
+      throw error
+    })
```
Comment (rfk) [inline test/local/mysql_tests.js]: Unnecessary `catch` here, left over from debugging maybe?

**Comment by rfk** [comment]
After a bit of mucking around in fxa-dev, this *appears* to be working as intended.  Here's what I did:

* Deployed an fxa-dev box with mysql strict-mode disabled, and auth-server modified to allow emoji in device names.
* Connected a device and set its name to "ryan's 🚲 Nightly".
* Confirmed that this configuration resulted in truncating device names at the emoji, showing the name as "ryan's " in the devices view.
* Updated the box to use this branch of auth-db-mysql, and confirmed that the migration had applied successfully.
* Edited the device name to "ryan 😍 Nightly" and saved it.
* Checked what had been written to the `name` and `nameUtf8` columns.

In the MySQL shell the value appears to have been corrupted in both places:

```
mysql> select name, nameUtf8 from devices;
+----------------+----------------+
| name           | nameUtf8       |
+----------------+----------------+
| ryan ? Nightly | ryan ? Nightly |
+----------------+----------------+
```

But as Vlad saw above, I think this is a client issue, because the hex values of the two columns are different:

```
mysql> select hex(name), hex(nameUtf8) from devices;
+------------------------------+------------------------------------+
| hex(name)                    | hex(nameUtf8)                      |
+------------------------------+------------------------------------+
| 7279616E203F204E696768746C79 | 7279616E20F09F988D204E696768746C79 |
+------------------------------+------------------------------------+
1 row in set (0.00 sec)
```

The first one gives the emoji the "?" treatment, as we might expect if it's getting mangled by the encoding:

```
$ python -c "print '7279616E203F204E696768746C79'.decode('hex')"
ryan ? Nightly
```

But the `nameUtf8` column appears to have the right bytes:

```
$ python -c "print '7279616E20F09F988D204E696768746C79'.decode('hex')"
ryan 😍 Nightly
```

Hopefully it will behave similarly in production :-)

(I had to test it using emoji rather than CJK characters, because the fxa-dev boxes create their db with utf8 by default, and hence do not exhibit https://bugzilla.mozilla.org/show_bug.cgi?id=1348298)
